file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/FileSystemConsumer.java

```diff
@@ -0,0 +1,672 @@
+            boolean isFileReadable;
+            try {
+                isFileExists = fileObject.exists();
+                isFileReadable = fileObject.isReadable();
+            } catch (FileSystemException e) {
+                throw new FileSystemServerConnectorException("Error occurred when determining" +
+                                                             " whether the file at URI : " +
+                                                             FileTransportUtils.maskURLPassword(fileURI) +
+                                                             " exists and readable. " + e);
+            }
+
+            if (isFileExists && isFileReadable) {
+                FileObject[] children = null;
+                try {
+                    children = fileObject.getChildren();
+                } catch (FileSystemException ignored) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("The file does not exist, or is not a folder, or an error " +
+                                  "has occurred when trying to list the children. File URI : " +
+                                  FileTransportUtils.maskURLPassword(fileURI), ignored);
+                    }
+                }
+                if (children == null || children.length == 0) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Folder at " + FileTransportUtils.maskURLPassword(fileURI) + " is empty.");
+                    }
+                } else {
+                    directoryHandler(children);
+                }
+            } else {
+                throw new FileSystemServerConnectorException(
+                        "Unable to access or read file or directory : " + FileTransportUtils.maskURLPassword(fileURI) +
+                        ". Reason: " + (isFileExists ? "The file can not be read!" : "The file does not exist!"));
+            }
+        } finally {
+            try {
+                fileObject.close();
+            } catch (FileSystemException e) {
+                log.warn("Could not close file at URI: " + FileTransportUtils.maskURLPassword(fileURI), e);
+            }
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("End : Scanning directory or file : " + FileTransportUtils.maskURLPassword(fileURI));
+        }
+    }
+
+    /**
+     * Handle directory with child elements.
+     *
+     * @param children The array containing child elements of a folder
+     */
+    private void directoryHandler(FileObject[] children) throws FileSystemServerConnectorException {
+        // Sort the files according to given properties
+        String strSortParam = fileProperties.get(Constants.FILE_SORT_PARAM);
+
+        if (strSortParam != null && !"NONE".equals(strSortParam)) {
+            log.debug("Starting to sort the files in folder: " + FileTransportUtils.maskURLPassword(fileURI));
+
+            String strSortOrder = fileProperties.get(Constants.FILE_SORT_ORDER);
```

Comment:
Sorting should be done in different method.

This breaks rule that a give method should only do one thing. This method has by products.